### DISCUSSION TITLE
Beginners luck running out - help needed accessing values.

### DISCUSSION THREAD

**Discussion by gazzatav** [taocpp/json#102] [Q&A] [answered]
Thanks for your work in making this library.  I needed to parse a response from a web api, installed taocpp-json from Conan, read some documentation and accessed the data I needed no problems.  Then I decided to experiment a bit more but when I try to use the 'to' member function I get the error in the commented code below.  I grepped the whole json repo for 'to<' and 'to <' but found no examples.  In the code I also comment with questions where I don't think I'm using the library correctly.  I'd appreciate feedback.

```
#include <iostream>
#include <string>
#include <tao/json.hpp>

int main(int argc, char const* argv[])
{
  if(argc < 2){
    std::cout << "Usage: expt2 example.json\n";
    return 1;
  }
  std::string filename(argv[1]);
  const tao::json::value v = tao::json::parse_file( filename );
  
// Why is meta_iter valid as a pointer (can compare to nullptr) but ver_iter is not?  Thought value v would be a map (like meta).
  auto meta_iter = v.find("meta");
  if(meta_iter == nullptr){
    std::cout << "No pointer\n";
    return 1;
  }
    
  auto meta = meta_iter -> get_object();
  auto ver_iter = meta.find("version");
  if(ver_iter == meta.end()){
    std::cerr << "Key 'version' not found.\n";
  }else
// Should I be using the library differently here? Using 'ver_iter -> first' feels like I'm not using it properly.
    if(ver_iter->first == "version"){
      std::string version;
      version = meta.at("version").as<std::string>();
// Why does this not work?
      //meta.at("version").to<std::string(version);
// error: ‘std::map<std::__cxx11::basic_string<char>, tao::json::basic_value<tao::json::traits>,
// std::less<std::__cxx11::basic_string<char> >, std::allocator<std::pair<const std::__cxx11::basic_string<char>, 
// tao::json::basic_value<tao::json::traits> > > >::mapped_type {aka class tao::json::basic_value<tao::json::traits>}’ 
// has no member named ‘to’
      std::cout << "Version: " << ver_iter -> second << ' ' << version << '\n';
    }
}
```

I've uploaded the json example response but had to name it example.txt as json uploads are not supported - the irony!
Thanks for any help.
[example.txt](https://github.com/taocpp/json/files/6528925/example.txt)

**Comment by ColinH** [comment, marked as answer]
> Why is `meta_iter` valid as a pointer (can compare to `nullptr`) but `ver_iter` is not?  Thought value `v` would be a map (like `meta`).

While `meta` is indeed a `std::map< std::string, tao::json::value >`, so the iterators are the usual map iterators, the problem is that `v` is a `tao::json::value` which can hold an object (`map`), an array (`vector`), or one of the many other values, so there is no clear candidate for what `tao::json::value::iterator` should be, and how it should behave.

For this reason the various `tao::json::value::find()` functions return a pointer to a `tao::json::value` that is `nullptr` when no appropriate sub-value was found (for objects the search key must be a string, for arrays the search key must be an integer).

**Comment by ColinH** [comment]
> Should I be using the library differently here? Using 'ver_iter -> first' feels like I'm not using it properly.

Checking whether `ver_iter->first` is equal to `"version"` is superfluous since `ver_iter` was the result of `find`ing `"version"`.

**Reply by gazzatav** [reply]
@ColinH Sorry, I should have explained what I was doing there much more clearly.  I was thinking how could I walk through some json to find a key at an unknown level, or even to find whether it was present, so I would need a way of testing for that key to see if I had found it.  Using ver_iter->first felt too much like descending into implementation details.

So from what you said above 'no clear candidate for what tao::json::value::iterator should be'  the proper way would be to use 'find' with the key at each level detecting the types of objects along the way.

**Reply by ColinH** [reply]
Yes, sounds like you need a custom find function that recursively enters arrays and objects to look for the given key "anywhere". If you could show us what you come up with we might consider adding it to `contrib`.

**Comment by ColinH** [comment]
> Why does this not work?

After adding the missing `>` for the template parameter it compiles for me...

```
meta.at("version").to<std::string>(version);
```

**Reply by gazzatav** [reply]
@ColinH   Thanks for that, I focussed on the compiler error instead of the typo.  I had tried previously to use 'to' to convert "version" to a double intead of a string like object so by the time I realized that mistake I was in a frame of mind where I thought I was doing something much more fundamentally wrong.

However after adding the missing '>' mine on g++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0 still does not compile with error:
```
expt2.cpp: In function ‘int main(int, const char**)’:
expt2.cpp:31:26: error: ‘std::map<std::__cxx11::basic_string<char>, tao::json::basic_value<tao::json::traits>, std::less<std::__cxx11::basic_string<char> >, std::allocator<std::pair<const std::__cxx11::basic_string<char>, tao::json::basic_value<tao::json::traits> > > >::mapped_type {aka class tao::json::basic_value<tao::json::traits>}’ has no member named ‘to’
       meta.at("version").to<std::string>(version);
                          ^~
expt2.cpp:31:40: error: expected primary-expression before ‘>’ token
       meta.at("version").to<std::string>(version);
```
